gph/src/ripser.h

```diff
@@ -691,9 +691,32 @@ class ripser
         };
 
         const index_t cnt = n - (k - 1);
+
         if (pred(n) || (cnt <= 0))
             return n;
-        return get_max(n, cnt, pred);
+
+        if (k != 3) {
+            return get_max(n, cnt, pred);
+        } else {
+            long double to_cbrt = 6 * idx;
```

Comment:
`long double` because indices are large here.

Reply:
`idx` is of type `index_t` which is of type `int64_t`.
Is there a reason that we convert to double ?
IMO, we could remain in as an integer, but If we are sure we do not overflow, we could go with `uint64_t`, but if it can overflow, we can let `index_t` and verify that `to_cbrt` is not `<0`.

Reply:
We need to take cube roots of a huge number, so a floating-point type is necessary, right?

Reply:
Yeah but the thing is that with double it seems that you can be less precise. See [link](https://ifnotnil.com/t/why-is-double-larger-than-int64/255).
Here also we are only doing a multiplication by an integer, this is a reason why I would keep it all integer, even if we can go a higher with double. Because for those big numbers, we will have less precision in double because we only work with the integer part.

Reply:
I'm confused about this, and particularly the implications for taking the square and cube roots.  Should we just keep the types what they are and hope that `std::sqrt` and `std::cbrt` deal with integer types in the best way possible?

Reply:
That is a good question, when looking at the documentation of both function [sqrt](https://en.cppreference.com/w/cpp/numeric/math/sqrt) and [cbrt](https://en.cppreference.com/w/cpp/numeric/math/cbrt), they both have a special input for any integer types, and if you look at a note, it indicates the following:

> A set of overloads or a function template accepting an argument of any integral type. Equivalent to 2) (the argument is cast to double).

It seems that when you input an integral type, it is implicitly casted to `double`.

I am not a big fan of `long double` because it is architecture dependent on the results provided. If we let `long double` 2 different users with the same input could lead to different results, because they have a different architecture, it will be a quite hard behavior to debug.

See [link](https://stackoverflow.com/a/14221894) for a detailed explanation on the format differences.

Reply:
Ok!  So it seems we are looking for a fast, exact implementation of the "integer square root" and "integer cube root".  I'm sure someone has done this!

Reply:
https://stackoverflow.com/a/31120562

Reply:
Nice, but we will need to look and test carefully this kind of solution, but I like that :smile:  !
I can have a look in my book "Hacker's delight" if there is a suggestion for this kind of misc functions.

Reply:
I can't think of a better person than you to help me figure this out!

Reply:
Thank you, will try to do my best with my current availability ...

Reply:
Boost! http://rosettacode.org/wiki/Isqrt_(integer_square_root)_of_X#C.2B.2B

Reply:
Other sources:
  - https://codereview.stackexchange.com/a/75345 (integer square root)
  - https://stackoverflow.com/a/4331882 (integer cube root)

Reply:
@MonkeyBreaker I have done a preliminary test and replacing `long double` with just `double` does seem to yield better performance.  So maybe another argument to remove `long double`.

Reply:
Well, for me it was not related to performances, but to inconsistent behavior with the same input data when run on 2 different architectures.
I would take with a grain of salt any performance difference between `double` and `long double`

Reply:
Right. For 2-simplices we cannot ensure the numerics will work exactly even for relatively small numbers, unlike the case of 1-simplices. So my point is that if using `double` still implies we are only 1 or two steps away from the correct results, and if it can be faster on some architectures, then it should be preferred.

Reply:
> if using double still implies we are only 1 or two steps away from the correct results for a wide range of indices

And I think this is true at least for datasets with as many as one million points.